reconcile/vault_replication.py

```diff
@@ -0,0 +1,176 @@
+import logging
+import re
+
+from reconcile.utils.vault import VaultClient, SecretNotFound
+from reconcile.gql_definitions.jenkins_configs import jenkins_configs
+
+from reconcile.gql_definitions.vault_policies import vault_policies
+from reconcile.gql_definitions.vault_instances import vault_instances
+from reconcile.gql_definitions.vault_instances.vault_instances import (
+    VaultReplicationV1,
+)
+from reconcile.gql_definitions.vault_policies.vault_policies import (
+    VaultPolicyV1,
+)
+
+from reconcile.utils import gql
+
+from typing import List
+
+QONTRACT_INTEGRATION = "vault-replication"
+
+
+def copy_vault_secret(
+    dry_run: bool, source_vault: VaultClient, dest_vault: VaultClient, path: str
+) -> None:
+
+    secret_dict = {"path": path, "version": "LATEST"}
+    secrets, version = source_vault.read_all_version(secret_dict)
+    for secret in secrets:
+        write_dict = {"path": path, "data": secret}
+        try:
+            _, dest_version = dest_vault.read_all(secret_dict)
+            if dest_version < version:
```

Comment:
consider this. both source and dest vault are in sync and both contain the secret version 3 for a certain path.
then the secret gets updated twice in source vault:
* version 4 has content x
* version 5 has content y

then the reconcile loop starts, detects that the source secret version is newer than the dest secret version and updates the secret, resulting in `version 4 with content y` in dest vault.

then the reconcile loop starts again, still detecting that the version in source is newer than in dest and updates the dest secret to `version 5 with content y`

at this point we have in source
* version 1,2,3
* version 4 with content x
* version 5 with content y

and in dest
* version 1,2,3
* version 4 with content y
* version 5 with content y

depending on how we want to use synced secrets this might be a concern or not

Reply:
When I was writing it I didn't think about it, but makes complete sense. I did a refactor to sync all the versions of a secret in any of the cases. With this we will have a complete mirror between vaults.